pkg/config/config_test.go

```diff
@@ -4,64 +4,119 @@ import (
 	"bytes"
 	"testing"
 
-	. "github.com/onsi/ginkgo"
-	. "github.com/onsi/gomega"
+	"github.com/stretchr/testify/require"
 )
 
-var environmentList = "environments:\n  - name: dev\n    auto: true\n"
+func TestConfigSimple(t *testing.T) {
+	data := `
+    environments:
+      - name: dev
+        auto: true
+      - name: qa
+        auto: true
+      - name: prod
+        auto: false
+  `
+	reader := bytes.NewReader([]byte(data))
+	cfg, err := LoadConfig(reader)
+	require.NoError(t, err)
 
-func TestNewGitHubGITProvider(t *testing.T) {
-	RegisterFailHandler(Fail)
-	RunSpecs(t, "Config")
-}
+	require.Len(t, cfg.Environments, 3)
+	require.Equal(t, PRFlowTypePerApp, cfg.PRFlow)
+	require.True(t, cfg.IsAnyEnvironmentManual())
 
-var _ = Describe("Config", func() {
-	var err error
-	var config Config
-	var configData string = "{}"
+	require.True(t, cfg.HasNextEnvironment("dev"))
+	require.True(t, cfg.HasNextEnvironment("qa"))
+	require.False(t, cfg.HasNextEnvironment("prod"))
 
-	JustBeforeEach(func() {
-		reader := bytes.NewReader([]byte(configData))
-		config, err = LoadConfig(reader)
-	})
+	_, err = cfg.IsEnvironmentAutomated("foobar")
+	require.EqualError(t, err, "environment named foobar does not exist")
+	automated, err := cfg.IsEnvironmentAutomated("dev")
+	require.NoError(t, err)
+	require.True(t, automated)
+	automated, err = cfg.IsEnvironmentAutomated("qa")
+	require.NoError(t, err)
+	require.True(t, automated)
+	automated, err = cfg.IsEnvironmentAutomated("prod")
+	require.NoError(t, err)
+	require.False(t, automated)
 
-	It("returns an error when mandatory values are missing", func() {
-		Expect(err).NotTo(BeNil())
-	})
+	_, err = cfg.NextEnvironment("foobar")
+	require.EqualError(t, err, "environment named foobar does not exist")
+	next, err := cfg.NextEnvironment("dev")
+	require.NoError(t, err)
+	require.Equal(t, "qa", next.Name)
+	next, err = cfg.NextEnvironment("qa")
+	require.NoError(t, err)
+	require.Equal(t, "prod", next.Name)
+	next, err = cfg.NextEnvironment("prod")
+	require.EqualError(t, err, "last environment cannot have a next environment")
 
-	Describe("With valid environments list", func() {
-		BeforeEach(func() {
-			configData = environmentList
-		})
+	_, err = cfg.PrevEnvironment("foobar")
+	require.EqualError(t, err, "environment named foobar does not exist")
+	prev, err := cfg.PrevEnvironment("prod")
+	require.NoError(t, err)
+	require.Equal(t, "qa", prev.Name)
+	prev, err = cfg.PrevEnvironment("qa")
+	require.NoError(t, err)
+	require.Equal(t, "dev", prev.Name)
+	prev, err = cfg.PrevEnvironment("dev")
+	require.EqualError(t, err, "first environment cannot have a previous environment")
+}
```

Comment:
I find this test hard to read and I don't see the contracts that these methods should uphold. In its current form, it will not really help in a refactoring, because it will only show its first failure on any run, forcing me to perform my refactoring in its entirety before I can be sure that I'm on the right track. Can you break it into separate tests? For example, it might make sense to have one test per environment or per use case.

One of the benefits with the style you replaced is that it makes clear the connection between input and output. It is basically a recursion on the question "If this one thing changes in the input, what happens to the output?" By contrast, this test takes a single input. An off-by-one error could easily result in the first or last env always being reported as auto or manual which this code will not see.

To be clear, the code under test is simple does not merit too intrusive an interrogation, but it is very hard to predict what a future reader will find hard to understand, so I think we want to be a little more explicit than this.

Reply:
Sure splitting this into separate tests would not be a problem. I do however not agree with you that this becomes harder to understand what contract are being enforced. Most of the logic of the tests were split into separate tests which basically do the same thing.

These additional tests were added to verify that the next previous logic was working. Which turned out that they did not work, and there were possible null pointer exceptions present. I am not a big fan of the ginkgo/gomega style of testing as it deviates a lot from "normal" golang syntax.

Reply:
I'm fine with this split-up.